File: src/main/java/org/voegtle/weatherwidget/util/ActivityUpdateTask.java
package org.voegtle.weatherwidget.util;

import android.os.AsyncTask;
import android.util.Log;
import android.widget.LinearLayout;
import org.voegtle.weatherwidget.R;
import org.voegtle.weatherwidget.WeatherActivity;
import org.voegtle.weatherwidget.data.WeatherData;
import org.voegtle.weatherwidget.location.LocationContainer;
import org.voegtle.weatherwidget.location.LocationIdentifier;
import org.voegtle.weatherwidget.location.LocationView;
import org.voegtle.weatherwidget.location.WeatherLocation;
import org.voegtle.weatherwidget.notification.NotificationSystemManager;
import org.voegtle.weatherwidget.preferences.ApplicationSettings;
import org.voegtle.weatherwidget.widget.ScreenPainterFactory;
import org.voegtle.weatherwidget.widget.WidgetScreenPainter;

import java.util.ArrayList;
import java.util.HashMap;

public class ActivityUpdateTask extends AsyncTask<Void, Void, HashMap<LocationIdentifier, WeatherData>> {
  private final WeatherDataFetcher weatherDataFetcher;
  private WeatherActivity activity;
  private boolean showToast;
  private ApplicationSettings configuration;

  ActivityUpdateTask(WeatherActivity activity, ApplicationSettings configuration, boolean showToast) {
    this.configuration = configuration;
    this.activity = activity;

    this.showToast = showToast;
    this.weatherDataFetcher = new WeatherDataFetcher(ContextUtil.getBuildNumber(activity));
  }


  @Override
  protected HashMap<LocationIdentifier, WeatherData> doInBackground(Void... voids) {
    return weatherDataFetcher.fetchAllWeatherDataFromServer(configuration.getLocations(), configuration.getSecret());
  }

  @Override
  protected void onPostExecute(HashMap<LocationIdentifier, WeatherData> data) {
    try {
      updateViewData(data);
      sortViews(data);
      updateWidgets(data);

      new UserFeedback(activity).showMessage(R.string.message_data_updated, showToast);

      NotificationSystemManager notificationManager = new NotificationSystemManager(activity, configuration);
      notificationManager.checkDataForAlert(data);
    } catch (Throwable th) {
      new UserFeedback(activity).showMessage(R.string.message_data_update_failed, true);
      Log.e(ActivityUpdateTask.class.toString(), "Failed to update View", th);
    }
  }

  private void updateWidgets(HashMap<LocationIdentifier, WeatherData> data) {
    ScreenPainterFactory factory = new ScreenPainterFactory(activity, configuration);
    ArrayList<WidgetScreenPainter> screenPainters = factory.createScreenPainters();
    for (WidgetScreenPainter screenPainter : screenPainters) {
      screenPainter.updateWidgetData(data);
      screenPainter.showDataIsValid();
    }
  }

  private void updateViewData(HashMap<LocationIdentifier, WeatherData> data) {
    for (WeatherLocation location : configuration.getLocations()) {
      WeatherData locationData = data.get(location.getKey());
      if (locationData != null) {
        updateWeatherLocation(location.getWeatherViewId(),
            location.getName(), locationData);
      }
    }
  }

  private void updateWeatherLocation(int locationId, String locationName, WeatherData data) {
    final LocationView contentView = (LocationView) activity.findViewById(locationId);

    final int color = ColorUtil.byAge(configuration.getColorScheme(), data.getTimestamp());
    final String caption = getCaption(locationName, data);

    updateView(contentView, caption, data, color);
  }

  private String getCaption(String locationName, WeatherData data) {
    return locationName + " - " + data.getLocaltime();
  }

  private void updateView(final LocationView view, final String caption, final WeatherData data, final int color) {
    view.setCaption(caption);
    view.setData(data);
    view.setTextColor(color);
  }

  private void sortViews(HashMap<LocationIdentifier, WeatherData> data) {
    LinearLayout container = (LinearLayout) (activity.findViewById(R.id.location_container));
    LocationContainer locationContainer = new LocationContainer(activity.getApplicationContext(), container, configuration);
    locationContainer.updateLocationOrder(data);
  }

}

File: src/main/java/org/voegtle/weatherwidget/util/ActivityUpdateTask.kt
package org.voegtle.weatherwidget.util

import android.os.AsyncTask
import android.util.Log
import kotlinx.android.synthetic.main.activity_weather.*
import org.voegtle.weatherwidget.R
import org.voegtle.weatherwidget.WeatherActivity
import org.voegtle.weatherwidget.data.WeatherData
import org.voegtle.weatherwidget.location.LocationContainer
import org.voegtle.weatherwidget.location.LocationIdentifier
import org.voegtle.weatherwidget.location.LocationOrderStore
import org.voegtle.weatherwidget.location.LocationView
import org.voegtle.weatherwidget.location.UserLocationUpdater
import org.voegtle.weatherwidget.location.WeatherLocation
import org.voegtle.weatherwidget.notification.NotificationSystemManager
import org.voegtle.weatherwidget.preferences.ApplicationSettings
import org.voegtle.weatherwidget.preferences.OrderCriteria
import org.voegtle.weatherwidget.widget.ScreenPainterFactory
import java.util.HashMap

class ActivityUpdateTask internal constructor(private val activity: WeatherActivity,
                                              private val configuration: ApplicationSettings,
                                              private val showToast: Boolean) : AsyncTask<Void, Void, FetchAllResponse>() {
  private val weatherDataFetcher = WeatherDataFetcher(ContextUtil.getBuildNumber(activity))
  private val userLocationUpdater = UserLocationUpdater(activity.applicationContext)
  private val locationOrderStore = LocationOrderStore(activity.applicationContext)
  private val formatter = DataFormatter()

  override fun doInBackground(vararg voids: Void): FetchAllResponse {
    userLocationUpdater.updateLocation()
    return weatherDataFetcher.fetchAllWeatherDataFromServer(configuration.locations, configuration.secret!!)
  }

  override fun onPostExecute(response: FetchAllResponse) {
    try {
      updateViewData(response.weatherMap)
      sortViews(response.weatherMap)
      updateWidgets(response.weatherMap)

      UserFeedback(activity).showMessage(
          if (response.valid) R.string.message_data_updated else R.string.message_data_update_failed, showToast)

      val notificationManager = NotificationSystemManager(activity, configuration)
      notificationManager.checkDataForAlert(response)
    } catch (th: Throwable) {
      UserFeedback(activity).showMessage(R.string.message_data_update_failed, true)
      Log.e(ActivityUpdateTask::class.java.toString(), "Failed to update View", th)
    }

  }

  private fun updateWidgets(data: HashMap<LocationIdentifier, WeatherData>) {
    val factory = ScreenPainterFactory(activity, configuration)
    val screenPainters = factory.createScreenPainters()
    for (screenPainter in screenPainters) {
      screenPainter.updateWidgetData(data)
      screenPainter.showDataIsValid()
    }
  }

  private fun updateViewData(data: HashMap<LocationIdentifier, WeatherData>) {
    configuration.locations.forEach { location ->
      data[location.key]?.let {
        updateWeatherLocation(location, location.name, it)
      }
    }
  }

  private fun updateWeatherLocation(location: WeatherLocation, locationName: String, data: WeatherData) {
    val contentView: LocationView = activity.findViewById(location.weatherViewId)

    val highlight = location.preferences.highlightActive
    val colorScheme = configuration.colorScheme
    val color = if (highlight && !DateUtil.isOutdated(data.timestamp))
      ColorUtil.highlightText(colorScheme)
    else
      ColorUtil.byAge(colorScheme, data.timestamp)

    val caption = getCaption(locationName, data)

    updateView(contentView, caption, data, color)
  }

  private fun getCaption(locationName: String, data: WeatherData): String {
    var caption = "$locationName - ${data.localtime}"

    if (locationOrderStore.readOrderCriteria() == OrderCriteria.location) {
      val userPosition = locationOrderStore.readPosition()
      val distance = userPosition.distanceTo(data.position)
      caption += " - ${formatter.formatDistance(distance.toFloat())}"
    }

    return caption
  }

  private fun updateView(view: LocationView, caption: String, data: WeatherData, color: Int) {
    view.setCaption(caption)
    view.setData(data)
    view.setTextColor(color)
  }

  private fun sortViews(data: HashMap<LocationIdentifier, WeatherData>) {
    val container = activity.location_container
    val locationContainer = LocationContainer(activity.applicationContext, container, configuration)
    locationContainer.updateLocationOrder(data)
  }

}
